nnfabrik/templates/transfer/transferred_trained_model.py

```diff
@@ -0,0 +1,310 @@
+        # {table_comment}
+        transfer_step:                     int          # transfer step
+        -> self().model_table
+        -> self().dataset_table
+        -> self().trainer_table
+        -> self().seed_table
+        collapsed_history:                 varchar(64)
+        data_transfer:                     tinyint 
+        ---
+        comment='':                        varchar(768) # short description 
+        score:                             float        # loss
+        output:                            longblob     # trainer object's output
+        ->[nullable] self().user_table
+        trainedmodel_ts=CURRENT_TIMESTAMP: timestamp    # UTZ timestamp at time of insertion
+        """.format(
+            table_comment=self.table_comment
+        )
+        return definition
+
+    class ModelStorage(TrainedModelBase.ModelStorage):
+        pass
+
+    class DataStorage(dj.Part):
+        @property
+        def definition(self):
+            definition = """
+            # Contains the data generated by the transfer step, stored externally.
+            -> master
+            ---
+            transfer_data:            attach@{storage}
+            """.format(
+                storage=self._master.storage
+            )
+            return definition
+
+    class CollapsedHistory(dj.Part):
+        """
+        For the result of two or more transfer steps to be uniquely identifiable,
+        we compress its entire history (the keys of all previous steps) into a single hash (`collapsed_history`).
+        This table keeps track of this process and can be used to recursively retrieve the transfer history.
+        """
+
+        definition = """
+        collapsed_history:                 varchar(64)
+        model_fn:                     varchar(64)
+        model_hash:                   varchar(64)
+        dataset_fn:                   varchar(64)
+        dataset_hash:                 varchar(64)
+        trainer_fn:                   varchar(64)
+        trainer_hash:                 varchar(64)
+        seed:                         int
+        prev_collapsed_history:             varchar(64)
+        """
+
+        @classmethod
+        def add_entry(cls, key):
+            key = {
+                "model_fn": key["model_fn"],
+                "model_hash": key["model_hash"],
+                "dataset_fn": key["dataset_fn"],
+                "dataset_hash": key["dataset_hash"],
+                "trainer_fn": key["trainer_fn"],
+                "trainer_hash": key["trainer_hash"],
+                "seed": key["seed"],
+                "prev_collapsed_history": key["collapsed_history"],
+            }
+            key["collapsed_history"] = make_hash(key)
+            cls.insert1(key, skip_duplicates=True)
+
+        def get_prev_key(self, key):
+            if key["collapsed_history"]:
+                prev_key = self & {"collapsed_history": key["collapsed_history"]}
+                return prev_key.proj(
```

Comment:
If this method is to stay here for the `CollapsedHistory` table, then this section will simplify the moment you change the definition of this table to use `collapsed_history` and `next_collapsed_history` as suggested above.